### PR TITLE
Lora sam

### PR DESCRIPTION
**Pull request by caroteu**

### CODE REVIEW

File: micro_sam/util.py [hunk 1]
```diff
@@ -347,7 +351,19 @@ def get_sam_model(
...
+
+    # change the state dict such that it matches sam's state dict
+    new_state_dict = model_state.copy()
+    for k, v in model_state.items():
+        if "image_encoder.blocks." in k and "attn.qkv.qkv" in k:
+            # Create new key with additional "qkv"
+            sam_key = k.replace(".qkv.qkv", ".qkv")
+            new_state_dict[sam_key] = new_state_dict.pop(k)
+        elif k not in list(sam.state_dict().keys()):
+            new_state_dict.pop(k)
```
Comment (anwai98) [inline micro_sam/util.py:363]: Hi @caroteu, could you explain this part, why do we do this for loading the model checkpoints?

Comment (caroteu) [reply micro_sam/util.py:363]: > Hi @caroteu, could you explain this part, why do we do this for loading the model checkpoints?

That is the part that is only necessary if the model was trained with lora. In order for sam.load_state_dict(model_state) to not throw an error, the state dict from the checkpoint needs to match the expected format of the sam state dict. Specifically the lora trained models have additional keys for the matrices A_q, B_q, A_v, B_v and the attention matrix weights are saved under a different name

Comment (anwai98) [reply micro_sam/util.py:363]: Just to get it straight, we match the LoRA model parameters with the SAM parameters in this stage and remove the mismatching ones?

Comment (caroteu) [reply micro_sam/util.py:363]: Yes

Comment (caroteu) [reply micro_sam/util.py:363]: Ok I see, then I rerun the experiment with that method and check if the results match the expectation

Comment (anwai98) [reply micro_sam/util.py:363]: I am running a round of experiments from my side. Let's compare the numbers and take it from there. I'll follow up on the numbers via RocketChat.

**Comment by caroteu** [comment]
This is outdated, can be closed

**Pull request closed without merging**